pkg/wrapper/buffered_tree.go

```diff
@@ -0,0 +1,215 @@
+func (p *TreePoolProvider) GetTreePool(squareSize uint64) *TreePool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if pool, exists := p.pools[squareSize]; exists {
+		return pool
+	}
+
+	pool := NewTreePool(squareSize, p.poolSize, p.opts...)
+	p.pools[squareSize] = pool
+	return pool
+}
+
+// TreePool provides pool methods for creating tree constructors
+type TreePool struct {
+	squareSize uint64
+	opts       []nmt.Option
+	poolSize   int
+	treePool   *fixedTreePool
+}
+
+// NewTreePool creates a new TreePool with the specified configuration
+func NewTreePool(squareSize uint64, poolSize int, opts ...nmt.Option) *TreePool {
+	return &TreePool{
+		squareSize: squareSize,
+		opts:       opts,
+		poolSize:   poolSize,
+		treePool:   newFixedTreePool(poolSize, squareSize, opts),
+	}
+}
+
+// BufferSize returns the number of trees in the pool
+func (f *TreePool) BufferSize() int {
+	return f.poolSize
+}
+
+// SquareSize returns the square size this pool was configured for
+func (f *TreePool) SquareSize() uint64 {
+	return f.squareSize
+}
+
+// NewConstructor returns a tree constructor function that uses the pool
+func (f *TreePool) NewConstructor() rsmt2d.TreeConstructorFn {
+	return func(_ rsmt2d.Axis, axisIndex uint) rsmt2d.Tree {
+		tree := f.treePool.acquire()
+		tree.axisIndex = uint64(axisIndex)
+		tree.shareIndex = 0
+
+		// Reset the tree (but don't put leaves back to pool anymore)
+		tree.tree.Reset()
+
+		return tree
+	}
+}
+
+var _ rsmt2d.Tree = &bufferedTree{}
+
+// bufferedTree is a wrapper around NamespaceMerkleTree with buffer pooling support
+// for efficient memory management in high-throughput scenarios.
+type bufferedTree struct {
+	squareSize uint64 // note: this refers to the width of the original square before erasure-coded
+	options    []nmt.Option
+	tree       *nmt.NamespacedMerkleTree
+	// axisIndex is the index of the axis (row or column) that this tree is on. This is passed
+	// by rsmt2d and used to help determine which quadrant each leaf belongs to.
+	axisIndex uint64
+	// shareIndex is the index of the share in a row or column that is being
+	// pushed to the tree. It is expected to be in the range: 0 <= shareIndex <
+	// 2*squareSize. shareIndex is used to help determine which quadrant each
+	// leaf belongs to, along with keeping track of how many leaves have been
+	// added to the tree so far.
+	shareIndex uint64
+	pool       *fixedTreePool // reference to the pool this tree belongs to
+	buffer     []byte         // Pre-allocated buffer for share data
+}
+
+// newBufferedTree creates a new bufferedTree with pre-allocated buffer and pool reference
+func newBufferedTree(squareSize uint64, axisIndex uint, pool *fixedTreePool, options ...nmt.Option) *bufferedTree {
+	if squareSize == 0 {
+		panic("cannot create a bufferedTree of squareSize == 0")
+	}
+	if pool == nil {
+		panic("cannot create a bufferedTree of pool == nil")
```

Comment:
if the pool is nil, can we create it instead of panicking?

Reply:
I think it is ok to leave it like this, because the constructor is private, so it should never be constructed without a pool and if it is - this is a critical problem in our implementation and we will catch it in tests